### PR TITLE
Add LDAP mode for basic authentication

### PR DESCRIPTION
**Pull request by samanhappy**

### CODE REVIEW

File: pkg/filters/validator/ldap.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+		UseSSL:             spec.UseSSL,
+		SkipTLS:            spec.SkipTLS,
+		InsecureSkipVerify: spec.Insecure,
+		ServerName:         spec.ServerName,
+		ClientCertificates: spec.certificates,
+	}
+	return v
+}
+
+var fnAuthLDAP = func(v *LDAPValidator, username, password string) error {
```
Comment (localvar) [inline pkg/filters/validator/ldap.go:79]: why define this function as a variable?

File: pkg/filters/validator/basicauth.go [hunk 2]
```diff
@@ -313,6 +337,62 @@ func (euc *etcdUserCache) Match(username string, password string) bool {
...
+		return false
+	}
+
+	userdn := fmt.Sprintf("%s=%s,%s", luc.spec.UID, username, luc.spec.BaseDN)
+	if err := luc.client.Conn.Bind(userdn, password); err != nil {
+		logger.Warnf("failed to bind LDAP user %v", err)
+		return false
+	}
+
+	return true
```
Comment (localvar) [inline pkg/filters/validator/basicauth.go:379]: could this function be called concurrently?

Comment (samanhappy) [reply pkg/filters/validator/basicauth.go]: this quesion is very considerate, the `client.Connect()` will create a connection if not exists and then store it which can be used by other function like `Conn.Bind`, but the code `defer luc.client.Close()` will break it and could case concurrency problem, so I remove it.

Comment (localvar) [reply pkg/filters/validator/basicauth.go]: do we sure that `luc.client.Conn.Bind` is goroutine safe.

Comment (samanhappy) [reply pkg/filters/validator/basicauth.go]: it should be safe according to [https://github.com/go-ldap/ldap/blob/master/ldap_test.go#L198](https://github.com/go-ldap/ldap/blob/master/ldap_test.go#L198), and I also did a test which shows safe.

**Review by localvar — APPROVED** [review]

**Review by suchen-sci — APPROVED** [review]

**Pull request merged**